I rebased and resolved the conflicts.

Current Build failure seems unrelated. Resteasy classes are missing.
`
[ERROR] testAggregate  Time elapsed: 0 s  <<< ERROR!
java.lang.NoClassDefFoundError: org/jboss/resteasy/core/ResteasyContext
	at org.jboss.resteasy.context.ResteasyContextProvider.currentContext(ResteasyContextProvider.java:18)
	at io.smallrye.context.impl.CapturedContextState.<init>(CapturedContextState.java:21)
	at io.smallrye.context.SmallRyeContextManager.captureContext(SmallRyeContextManager.java:64)
	at io.smallrye.context.SmallRyeThreadContext.currentContextExecutor(SmallRyeThreadContext.java:208)`

--- a/docs/src/main/asciidoc/mongodb.adoc
+++ b/docs/src/main/asciidoc/mongodb.adoc
@@ -543,6 +543,14 @@ So when you access the `/health/ready` endpoint of your application you will hav
 
 This behavior can be disabled by setting the `quarkus.mongodb.health.enabled` property to `false` in your `application.properties`.
 
+== Metrics
+
+If you are using the `quarkus-smallrye-metrics` extension, `quarkus-mongodb` can provide metrics about the connection pools.
+This behavior must first be enabled by setting the `quarkus.mongodb.metrics.enabled` property to `true` in your `application.properties`.
+
+So when you access the `/metrics` endpoint of your application you will have information about the connection pool status in the `vendor` scope.
+
+
 == The legacy client
 
 We don't include the legacy MongoDB client by default. It contains the now retired MongoDB Java API (DB, DBCollection,... )

--- a/extensions/mongodb-client/deployment/pom.xml
+++ b/extensions/mongodb-client/deployment/pom.xml
@@ -40,6 +40,11 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-metrics</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5-internal</artifactId>
             <scope>test</scope>
         </dependency>

--- a/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java
+++ b/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java
@@ -11,4 +11,10 @@ public class MongoClientBuildTimeConfig {
      */
     @ConfigItem(name = "health.enabled", defaultValue = "true")
     public boolean healthEnabled;
+
+    /**
+     * Whether or not metrics are published in case the smallrye-metrics extension is present.
+     */
+    @ConfigItem(name = "metrics.enabled")
+    public boolean metricsEnabled;
 }

--- a/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java
+++ b/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java
@@ -24,6 +24,7 @@ import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
 
 import com.mongodb.client.MongoClient;
+import com.mongodb.event.ConnectionPoolListener;
 
 import io.quarkus.arc.Unremovable;
 import io.quarkus.arc.deployment.BeanContainerListenerBuildItem;
@@ -32,6 +33,7 @@ import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
 import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
 import io.quarkus.arc.processor.DotNames;
+import io.quarkus.deployment.Capabilities;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Record;
@@ -49,6 +51,7 @@ import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.mongodb.ReactiveMongoClient;
+import io.quarkus.mongodb.metrics.MongoMetricsConnectionPoolListener;
 import io.quarkus.mongodb.runtime.AbstractMongoClientProducer;
 import io.quarkus.mongodb.runtime.MongoClientConfig;
 import io.quarkus.mongodb.runtime.MongoClientName;
@@ -75,9 +78,16 @@ public class MongoClientProcessor {
     void configureRuntimeProperties(MongoClientRecorder recorder,
             CodecProviderBuildItem codecProvider,
             BsonDiscriminatorBuildItem bsonDiscriminator,
-            MongodbConfig config) {
+            MongodbConfig config,
+            List<MongoConnectionPoolListenerBuildItem> connectionPoolListenerProvider) {
+        List<ConnectionPoolListener> poolListenerList = connectionPoolListenerProvider.stream()
+                .map(MongoConnectionPoolListenerBuildItem::getConnectionPoolListener)
+                .collect(Collectors.toList());
+
         recorder.configureRuntimeProperties(codecProvider.getCodecProviderClassNames(),
-                bsonDiscriminator.getBsonDiscriminatorClassNames(), config);
+                bsonDiscriminator.getBsonDiscriminatorClassNames(),
+                config,
+                poolListenerList);
     }
 
     @BuildStep
@@ -391,5 +401,15 @@ public class MongoClientProcessor {
     HealthBuildItem addHealthCheck(MongoClientBuildTimeConfig buildTimeConfig) {
         return new HealthBuildItem("io.quarkus.mongodb.health.MongoHealthCheck",
                 buildTimeConfig.healthEnabled, "mongodb");
+    }
+
+    @BuildStep
+    void setupMetrics(
+            MongoClientBuildTimeConfig buildTimeConfig, Capabilities capabilities,
+            BuildProducer<MongoConnectionPoolListenerBuildItem> producer) {
+
+        if (buildTimeConfig.metricsEnabled && capabilities.isCapabilityPresent(Capabilities.METRICS)) {
+            producer.produce(new MongoConnectionPoolListenerBuildItem(new MongoMetricsConnectionPoolListener()));
+        }
     }
 }

--- a/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoConnectionPoolListenerBuildItem.java
+++ b/extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoConnectionPoolListenerBuildItem.java
@@ -1,0 +1,17 @@
+package io.quarkus.mongodb.deployment;
+
+import com.mongodb.event.ConnectionPoolListener;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class MongoConnectionPoolListenerBuildItem extends MultiBuildItem {
+    private ConnectionPoolListener connectionPoolListener;
+
+    public MongoConnectionPoolListenerBuildItem(ConnectionPoolListener connectionPoolListener) {
+        this.connectionPoolListener = connectionPoolListener;
+    }
+
+    public ConnectionPoolListener getConnectionPoolListener() {
+        return connectionPoolListener;
+    }
+}

--- a/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java
+++ b/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java
@@ -1,0 +1,77 @@
+package io.quarkus.mongodb;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.metrics.Metric;
+import org.eclipse.microprofile.metrics.MetricID;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.Tag;
+import org.eclipse.microprofile.metrics.annotation.RegistryType;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.mongodb.client.MongoClient;
+
+import io.quarkus.mongodb.metrics.ConnectionPoolGauge;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MongoMetricsTest extends MongoTestBase {
+
+    @Inject
+    MongoClient client;
+
+    @Inject
+    @RegistryType(type = MetricRegistry.Type.VENDOR)
+    MetricRegistry registry;
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(MongoTestBase.class))
+            .withConfigurationResource("application-metrics-mongo.properties");
+
+    @AfterEach
+    void cleanup() {
+        if (client != null) {
+            client.close();
+        }
+    }
+
+    @Test
+    void testMetricsInitialization() {
+        assertNull(getGaugeValueOrNull("mongodb.connection-pool.size", getTags()));
+        assertNull(getGaugeValueOrNull("mongodb.connection-pool.checked-out-count", getTags()));
+
+        // Just need to execute something so that an connection is opened
+        String name = client.listDatabaseNames().first();
+
+        assertEquals(1L, getGaugeValueOrNull("mongodb.connection-pool.size", getTags()));
+        assertEquals(0L, getGaugeValueOrNull("mongodb.connection-pool.checked-out-count", getTags()));
+
+        client.close();
+        assertEquals(0L, getGaugeValueOrNull("mongodb.connection-pool.size", getTags()));
+        assertEquals(0L, getGaugeValueOrNull("mongodb.connection-pool.checked-out-count", getTags()));
+    }
+
+    private Long getGaugeValueOrNull(String metricName, Tag[] tags) {
+        MetricID metricID = new MetricID(metricName, tags);
+        Metric metric = registry.getMetrics().get(metricID);
+
+        if (metric == null) {
+            return null;
+        }
+        return ((ConnectionPoolGauge) metric).getValue();
+    }
+
+    private Tag[] getTags() {
+        return new Tag[] {
+                new Tag("host", "localhost"),
+                new Tag("port", "27018"),
+        };
+    }
+}

--- a/extensions/mongodb-client/deployment/src/test/resources/application-metrics-mongo.properties
+++ b/extensions/mongodb-client/deployment/src/test/resources/application-metrics-mongo.properties
@@ -1,0 +1,2 @@
+quarkus.mongodb.connection-string=mongodb://localhost:27018
+quarkus.mongodb.metrics.enabled=true

--- a/extensions/mongodb-client/runtime/pom.xml
+++ b/extensions/mongodb-client/runtime/pom.xml
@@ -49,6 +49,12 @@
         </dependency>
 
         <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-smallrye-metrics</artifactId>
+            <optional>true</optional>
+        </dependency>
+
+        <dependency>
             <groupId>org.graalvm.nativeimage</groupId>
             <artifactId>svm</artifactId>
         </dependency>

--- a/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/ConnectionPoolGauge.java
+++ b/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/ConnectionPoolGauge.java
@@ -1,0 +1,22 @@
+package io.quarkus.mongodb.metrics;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.eclipse.microprofile.metrics.Gauge;
+
+public class ConnectionPoolGauge implements Gauge<Long> {
+    private AtomicLong value = new AtomicLong();
+
+    public void decrement() {
+        value.decrementAndGet();
+    }
+
+    public void increment() {
+        value.incrementAndGet();
+    }
+
+    @Override
+    public Long getValue() {
+        return value.longValue();
+    }
+}

--- a/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java
+++ b/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/metrics/MongoMetricsConnectionPoolListener.java
@@ -1,0 +1,117 @@
+package io.quarkus.mongodb.metrics;
+
+import org.eclipse.microprofile.metrics.Metadata;
+import org.eclipse.microprofile.metrics.Metric;
+import org.eclipse.microprofile.metrics.MetricID;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.MetricType;
+import org.eclipse.microprofile.metrics.Tag;
+
+import com.mongodb.connection.ServerId;
+import com.mongodb.event.ConnectionAddedEvent;
+import com.mongodb.event.ConnectionCheckedInEvent;
+import com.mongodb.event.ConnectionCheckedOutEvent;
+import com.mongodb.event.ConnectionPoolClosedEvent;
+import com.mongodb.event.ConnectionPoolListener;
+import com.mongodb.event.ConnectionPoolOpenedEvent;
+import com.mongodb.event.ConnectionPoolWaitQueueEnteredEvent;
+import com.mongodb.event.ConnectionPoolWaitQueueExitedEvent;
+import com.mongodb.event.ConnectionRemovedEvent;
+
+import io.smallrye.metrics.MetricRegistries;
+
+public class MongoMetricsConnectionPoolListener implements ConnectionPoolListener {
+    private final static String SIZE_NAME = "mongodb.connection-pool.size";
+    private final static String CHECKED_OUT_COUNT_NAME = "mongodb.connection-pool.checked-out-count";
+
+    @Override
+    public void connectionPoolOpened(ConnectionPoolOpenedEvent event) {
+        Tag[] tags = createTags(event.getServerId());
+
+        registerGauge(SIZE_NAME, "the current size of the pool, including idle and and in-use members", tags);
+        registerGauge(CHECKED_OUT_COUNT_NAME, "the current count of connections that are currently in use", tags);
+    }
+
+    @Override
+    public void connectionPoolClosed(ConnectionPoolClosedEvent event) {
+    }
+
+    @Override
+    public void connectionCheckedOut(ConnectionCheckedOutEvent event) {
+        MetricID metricID = createMetricID(CHECKED_OUT_COUNT_NAME, event.getConnectionId().getServerId());
+
+        Metric metric = getMetricRegistry().getMetrics().get(metricID);
+
+        if (metric != null) {
+            ((ConnectionPoolGauge) metric).increment();
+        }
+    }
+
+    @Override
+    public void connectionCheckedIn(ConnectionCheckedInEvent event) {
+        MetricID metricID = createMetricID(CHECKED_OUT_COUNT_NAME, event.getConnectionId().getServerId());
+
+        Metric metric = getMetricRegistry().getMetrics().get(metricID);
+
+        if (metric != null) {
+            ((ConnectionPoolGauge) metric).decrement();
+        }
+    }
+
+    @Override
+    public void waitQueueEntered(ConnectionPoolWaitQueueEnteredEvent connectionPoolWaitQueueEnteredEvent) {
+        // Not supported, since the event is deprecated and will be removed anyway
+    }
+
+    @Override
+    public void waitQueueExited(ConnectionPoolWaitQueueExitedEvent connectionPoolWaitQueueExitedEvent) {
+        // Not supported, since the event is deprecated and will be removed anyway
+    }
+
+    @Override
+    public void connectionAdded(ConnectionAddedEvent event) {
+
+        MetricID metricID = createMetricID(SIZE_NAME, event.getConnectionId().getServerId());
+
+        Metric metric = getMetricRegistry().getMetrics().get(metricID);
+
+        if (metric != null) {
+            ((ConnectionPoolGauge) metric).increment();
+        }
+    }
+
+    @Override
+    public void connectionRemoved(ConnectionRemovedEvent event) {
+
+        MetricID metricID = createMetricID(SIZE_NAME, event.getConnectionId().getServerId());
+
+        Metric metric = getMetricRegistry().getMetrics().get(metricID);
+
+        if (metric != null) {
+            ((ConnectionPoolGauge) metric).decrement();
+        }
+    }
+
+    private void registerGauge(String metricName, String description, Tag[] tags) {
+        getMetricRegistry().remove(new MetricID(metricName, tags));
+
+        Metadata metaData = Metadata.builder().withName(metricName).withType(MetricType.GAUGE)
+                .withDescription(description).build();
+        getMetricRegistry().register(metaData, new ConnectionPoolGauge(), tags);
+    }
+
+    private MetricRegistry getMetricRegistry() {
+        return MetricRegistries.get(MetricRegistry.Type.VENDOR);
+    }
+
+    private Tag[] createTags(ServerId server) {
+        return new Tag[] {
+                new Tag("host", server.getAddress().getHost()),
+                new Tag("port", String.valueOf(server.getAddress().getPort())),
+        };
+    }
+
+    private MetricID createMetricID(String metricName, ServerId server) {
+        return new MetricID(metricName, createTags(server));
+    }
+}

--- a/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/AbstractMongoClientProducer.java
+++ b/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/AbstractMongoClientProducer.java
@@ -44,6 +44,7 @@ import com.mongodb.connection.ConnectionPoolSettings;
 import com.mongodb.connection.ServerSettings;
 import com.mongodb.connection.SocketSettings;
 import com.mongodb.connection.SslSettings;
+import com.mongodb.event.ConnectionPoolListener;
 
 import io.quarkus.mongodb.impl.AxleReactiveMongoClientImpl;
 import io.quarkus.mongodb.impl.ReactiveMongoClientImpl;
@@ -57,6 +58,7 @@ public abstract class AbstractMongoClientProducer {
     private boolean disableSslSupport = false;
     private List<String> codecProviders;
     private List<String> bsonDiscriminators;
+    private List<ConnectionPoolListener> connectionPoolListeners;
     private Map<String, MongoClient> mongoclients = new HashMap<>();
     private Map<String, ReactiveMongoClient> reactiveMongoClients = new HashMap<>();
     private Map<String, io.quarkus.mongodb.ReactiveMongoClient> legacyReactiveMongoClients = new HashMap<>();
@@ -147,11 +149,13 @@ public abstract class AbstractMongoClientProducer {
     }
 
     private static class ConnectionPoolSettingsBuilder implements Block<ConnectionPoolSettings.Builder> {
-        public ConnectionPoolSettingsBuilder(MongoClientConfig config) {
+        public ConnectionPoolSettingsBuilder(MongoClientConfig config, List<ConnectionPoolListener> connectionPoolListeners) {
             this.config = config;
+            this.connectionPoolListeners = connectionPoolListeners;
         }
 
         private MongoClientConfig config;
+        private List<ConnectionPoolListener> connectionPoolListeners;
 
         @Override
         public void apply(ConnectionPoolSettings.Builder builder) {
@@ -169,6 +173,9 @@ public abstract class AbstractMongoClientProducer {
             }
             if (config.maintenanceInitialDelay.isPresent()) {
                 builder.maintenanceInitialDelay(config.maintenanceInitialDelay.get().toMillis(), TimeUnit.MILLISECONDS);
+            }
+            for (ConnectionPoolListener connectionPoolListener : connectionPoolListeners) {
+                builder.addConnectionPoolListener(connectionPoolListener);
             }
         }
     }
@@ -288,13 +295,14 @@ public abstract class AbstractMongoClientProducer {
             settings.applyToSslSettings(new SslSettingsBuilder(config, disableSslSupport));
         }
         settings.applyToClusterSettings(new ClusterSettingBuilder(config));
-        settings.applyToConnectionPoolSettings(new ConnectionPoolSettingsBuilder(config));
+        settings.applyToConnectionPoolSettings(new ConnectionPoolSettingsBuilder(config, connectionPoolListeners));
         settings.applyToServerSettings(new ServerSettingsBuilder(config));
         settings.applyToSocketSettings(new SocketSettingsBuilder(config));
 
         if (config.readPreference.isPresent()) {
             settings.readPreference(ReadPreference.valueOf(config.readPreference.get()));
         }
+
         return settings.build();
     }
 
@@ -402,6 +410,10 @@ public abstract class AbstractMongoClientProducer {
 
     public void setBsonDiscriminators(List<String> bsonDiscriminators) {
         this.bsonDiscriminators = bsonDiscriminators;
+    }
+
+    public void setConnectionPoolListeners(List<ConnectionPoolListener> connectionPoolListeners) {
+        this.connectionPoolListeners = connectionPoolListeners;
     }
 
     public void disableSslSupport() {

--- a/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java
+++ b/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java
@@ -3,6 +3,7 @@ package io.quarkus.mongodb.runtime;
 import java.util.List;
 
 import com.mongodb.client.MongoClient;
+import com.mongodb.event.ConnectionPoolListener;
 
 import io.quarkus.arc.Arc;
 import io.quarkus.arc.runtime.BeanContainer;
@@ -30,7 +31,8 @@ public class MongoClientRecorder {
         };
     }
 
-    public void configureRuntimeProperties(List<String> codecs, List<String> bsonDiscriminators, MongodbConfig config) {
+    public void configureRuntimeProperties(List<String> codecs, List<String> bsonDiscriminators, MongodbConfig config,
+            List<ConnectionPoolListener> connectionPoolListeners) {
         // TODO @dmlloyd
         // Same here, the map is entirely empty (obviously, I didn't expect the values
         // that were not properly injected but at least the config objects present in
@@ -40,6 +42,7 @@ public class MongoClientRecorder {
         producer.setCodecs(codecs);
         producer.setBsonDiscriminators(bsonDiscriminators);
         producer.setConfig(config);
+        producer.setConnectionPoolListeners(connectionPoolListeners);
     }
 
     public RuntimeValue<MongoClient> getClient(String name) {
@@ -51,5 +54,4 @@ public class MongoClientRecorder {
         AbstractMongoClientProducer producer = Arc.container().instance(AbstractMongoClientProducer.class).get();
         return new RuntimeValue<>(producer.getReactiveClient(name));
     }
-
 }

